File: medium/cpp/1315.cpp
// https://leetcode.com/problems/sum-of-nodes-with-even-valued-grandparent/description/

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution
{
public:
    int sumEvenGrandparent(TreeNode* root)
    {
        return solve(root, nullptr, nullptr);
    }
    
private:
    int solve(TreeNode* root, TreeNode* parent, TreeNode* grandparent)
    {
        int answer = 0;
        
        if (root == nullptr)
        {
            return answer;
        }
        
        if (grandparent != nullptr && grandparent->val % 2 == 0)
        {
            answer += root->val;
        }
        
        if (root->left != nullptr)
        {
            answer += solve(root->left, root, parent);
        }
        
        if (root->right != nullptr)
        {
            answer += solve(root->right, root, parent);
        }
        
        return answer;
    }
};

File: medium/python/1315.py
# https://leetcode.com/problems/sum-of-nodes-with-even-valued-grandparent/description/

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def sumEvenGrandparent(self, root: TreeNode) -> int:
        return self.solve(root)

    def solve(self, root, parent = None, gparent = None):
        if root is None:
            return 0
            
        return (0 if gparent is None or gparent.val % 2 == 1 else root.val) + \
            self.solve(root.left, root, parent) + \
            self.solve(root.right, root, parent)